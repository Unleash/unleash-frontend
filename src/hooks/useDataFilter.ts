import React, { useState } from 'react';

export const useDataFilter = (
    data: any,
    filters: string[],
    setter: React.SetStateAction<React.Dispatch<any>>
) => {
    // setup filter state
    // return out filter functions to use in UI
    // filter data

    const filter = (data: any) => {
        return data;
    };

    const changeFilterState = () => {
        // change state for filter
    };

    return {
        filteredData: filter(data),
        filters: [{ name: 'merchant', onClick: changeFilterState }],
    };
};
