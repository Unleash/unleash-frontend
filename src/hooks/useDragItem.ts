import { useRef, useEffect, RefObject } from 'react';

export type MoveListItem = (
    dragIndex: number,
    dropIndex: number,
    save?: boolean
) => void;

export const useDragItem = (
    listItemIndex: number,
    moveListItem: MoveListItem
): RefObject<HTMLLIElement> => {
    const ref = useRef<HTMLLIElement>(null);

    useEffect(() => {
        if (ref.current) {
            ref.current.draggable = true;
            ref.current.dataset.index = String(listItemIndex);
            return addEventListeners(ref.current, moveListItem);
        }
    }, [listItemIndex, moveListItem]);

    return ref;
};

const addEventListeners = (
    el: HTMLLIElement,
    moveListItem: MoveListItem
): (() => void) => {
    const moveDraggedElement = (save: boolean) => {
        if (globalDraggedElement) {
            moveListItem(
                Number(globalDraggedElement.dataset.index),
                Number(el.dataset.index),
                save
            );
        }
    };

    const onDragStart = () => {
        globalDraggedElement = el;
    };

    const onDragEnter = () => {
        moveDraggedElement(false);
    };

    const onDragOver = (event: DragEvent) => {
        event.preventDefault();
    };

    const onDrop = () => {
        moveDraggedElement(true);
        globalDraggedElement = null;
    };

    el.addEventListener('dragstart', onDragStart);
    el.addEventListener('dragenter', onDragEnter);
    el.addEventListener('dragover', onDragOver);
    el.addEventListener('drop', onDrop);

    return () => {
        el.removeEventListener('dragstart', onDragStart);
        el.removeEventListener('dragenter', onDragEnter);
        el.removeEventListener('dragover', onDragOver);
        el.removeEventListener('drop', onDrop);
    };
};

// The element being dragged in the browser.
let globalDraggedElement: HTMLLIElement | null;
