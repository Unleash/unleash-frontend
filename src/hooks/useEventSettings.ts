import { getBasePath } from '../utils/format-path';
import { createPersistentGlobalStateHook } from './usePersistentGlobalState';
import React from 'react';

export interface IEventSettings {
    showData: boolean;
}

interface IUseEventSettingsOutput {
    eventSettings: IEventSettings;
    setEventSettings: React.Dispatch<React.SetStateAction<IEventSettings>>;
}

export const useEventSettings = (): IUseEventSettingsOutput => {
    const [eventSettings, setEventSettings] = useGlobalState();

    return { eventSettings, setEventSettings };
};

const createInitialValue = (): IEventSettings => {
    return { showData: false };
};

const useGlobalState = createPersistentGlobalStateHook<IEventSettings>(
    `${getBasePath()}:useEventSettings:v1`,
    createInitialValue()
);
