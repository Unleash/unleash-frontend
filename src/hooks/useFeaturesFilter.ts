import { IFeatureToggle } from '../interfaces/featureToggle';
import React, { useMemo } from 'react';
import { getBasePath } from '../utils/format-path';
import { createPersistentGlobalState } from './usePersistentGlobalState';

export interface IFeaturesFilter {
    query?: string;
    project: string;
}

export interface IFeaturesSortOutput {
    filtered: IFeatureToggle[];
    filter: IFeaturesFilter;
    setFilter: React.Dispatch<React.SetStateAction<IFeaturesFilter>>
}

// Store the features filter state globally, and in localStorage.
// When changing the format of IFeaturesFilter, change the version as well.
const useFeaturesFilterState = createPersistentGlobalState<IFeaturesFilter>(
    `${getBasePath()}:useFeaturesFilter:v1`,
    { project: '*' }
);

export const useFeaturesFilter = (
    features: IFeatureToggle[]
): IFeaturesSortOutput => {
    const [filter, setFilter] = useFeaturesFilterState();

    const filtered = useMemo(() => {
        return filterFeatures(features, filter);
    }, [features, filter]);

    return {
        setFilter,
        filter,
        filtered,
    };
};

// Return the current project ID a project has been selected,
// or the 'default' project if showing all projects.
export const resolveFilteredProjectId = (filter: IFeaturesFilter): string => {
    if (!filter.project || filter.project === '*') {
        return 'default';
    }

    return filter.project;
};

const filterFeatures = (
    features: IFeatureToggle[],
    filter: IFeaturesFilter
): IFeatureToggle[] => {
    return filterFeaturesByQuery(
        filterFeaturesByProject(features, filter),
        filter
    );
};

const filterFeaturesByProject = (
    features: IFeatureToggle[],
    filter: IFeaturesFilter
): IFeatureToggle[] => {
    return filter.project === '*'
        ? features
        : features.filter(f => f.project === filter.project);
};

const filterFeaturesByQuery = (
    features: IFeatureToggle[],
    filter: IFeaturesFilter
): IFeatureToggle[] => {
    if (!filter.query) {
        return features;
    }

    // Try to parse the search query as a RegExp.
    // Return all features if it can't be parsed.
    try {
        const regExp = new RegExp(filter.query, 'i');
        return features.filter(f => filterFeatureByRegExp(f, filter, regExp));
    } catch (err) {
        if (err instanceof SyntaxError) {
            return features;
        } else {
            throw err;
        }
    }
};

const filterFeatureByRegExp = (
    feature: IFeatureToggle,
    filter: IFeaturesFilter,
    regExp: RegExp
): boolean => {
    if (regExp.test(feature.name) || regExp.test(feature.description)) {
        return true;
    }

    if (
        filter.query &&
        filter.query.length > 1 &&
        regExp.test(JSON.stringify(feature))
    ) {
        return true;
    }

    if (!feature.strategies) {
        return false;
    }

    return feature.strategies.some(
        s =>
            regExp.test(s.name) ||
            s.constraints.some(c => c.values.some(v => regExp.test(v)))
    );
};
