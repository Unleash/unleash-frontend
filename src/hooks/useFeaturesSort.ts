import { IFeatureToggle } from '../interfaces/featureToggle';
import React, { useMemo } from 'react';
import { getBasePath } from '../utils/format-path';
import { createPersistentGlobalState } from './usePersistentGlobalState';

type FeaturesSortType =
    | 'name'
    | 'type'
    | 'enabled'
    | 'stale'
    | 'created'
    | 'last-seen'
    | 'project';

interface IFeaturesSort {
    type: FeaturesSortType;
}

export interface IFeaturesSortOutput {
    sort: IFeaturesSort;
    sorted: IFeatureToggle[];
    setSort: React.Dispatch<React.SetStateAction<IFeaturesSort>>;
}

export interface IFeaturesFilterSortOption {
    type: FeaturesSortType;
    name: string;
}

// Store the features sort state globally, and in localStorage.
// When changing the format of IFeaturesSort, change the version as well.
const useFeaturesSortState = createPersistentGlobalState<IFeaturesSort>(
    `${getBasePath()}:useFeaturesSort:v1`,
    { type: 'name' }
);

export const useFeaturesSort = (
    features: IFeatureToggle[]
): IFeaturesSortOutput => {
    const [sort, setSort] = useFeaturesSortState();

    const sorted = useMemo(() => {
        return sortFeatures(features, sort);
    }, [features, sort]);

    return {
        setSort,
        sort,
        sorted,
    };
};

export const createFeaturesFilterSortOptions =
    (): IFeaturesFilterSortOption[] => {
        return [
            { type: 'name', name: 'Name' },
            { type: 'type', name: 'Type' },
            { type: 'enabled', name: 'Enabled' },
            { type: 'stale', name: 'Stale' },
            { type: 'created', name: 'Created' },
            { type: 'last-seen', name: 'Last seen' },
            { type: 'project', name: 'Project' },
        ];
    };

const sortFeatures = (
    features: IFeatureToggle[],
    sort: IFeaturesSort
): IFeatureToggle[] => {
    switch (sort.type) {
        case 'enabled':
            return sortByEnabled(features);
        case 'stale':
            return sortByStale(features);
        case 'created':
            return sortByCreated(features);
        case 'last-seen':
            return sortByLastSeen(features);
        case 'name':
            return sortByName(features);
        case 'project':
            return sortByProject(features);
        case 'type':
            return sortByType(features);
        default:
            console.error(`Unknown feature sort type: ${sort.type}`);
            return features;
    }
};

const sortByEnabled = (
    features: Readonly<IFeatureToggle[]>
): IFeatureToggle[] => {
    return [...features].sort((a, b) =>
        a.enabled === b.enabled ? 0 : a.enabled ? -1 : 1
    );
};

const sortByStale = (
    features: Readonly<IFeatureToggle[]>
): IFeatureToggle[] => {
    return [...features].sort((a, b) =>
        a.stale === b.stale ? 0 : a.stale ? -1 : 1
    );
};

const sortByLastSeen = (
    features: Readonly<IFeatureToggle[]>
): IFeatureToggle[] => {
    return [...features].sort((a, b) =>
        a.lastSeenAt && b.lastSeenAt
            ? a.lastSeenAt.localeCompare(b.lastSeenAt)
            : 0
    );
};

const sortByCreated = (
    features: Readonly<IFeatureToggle[]>
): IFeatureToggle[] => {
    return [...features].sort((a, b) =>
        new Date(a.createdAt) > new Date(b.createdAt) ? -1 : 1
    );
};

const sortByName = (features: Readonly<IFeatureToggle[]>): IFeatureToggle[] => {
    return [...features].sort((a, b) => a.name.localeCompare(b.name));
};

const sortByProject = (
    features: Readonly<IFeatureToggle[]>
): IFeatureToggle[] => {
    return [...features].sort((a, b) => a.project.localeCompare(b.project));
};

const sortByType = (features: Readonly<IFeatureToggle[]>): IFeatureToggle[] => {
    return [...features].sort((a, b) => a.type.localeCompare(b.type));
};
