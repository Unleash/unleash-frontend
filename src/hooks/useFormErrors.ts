import { useState, useCallback, useMemo } from 'react';
import produce from 'immer';

export interface IFormErrors {
    // Get the error message for a field name, if any.
    get(field: string): string | undefined;

    // Set an error message for a field name.
    set(field: string, message: string): void;

    // Remove an existing error for a field name.
    remove(field: string): void;

    // Check if there are any errors.
    some(): boolean;
}

export const useFormErrors = (): IFormErrors => {
    const [errors, setErrors] = useState<Record<string, string>>({});

    const get = useCallback(
        (field: string): string | undefined => errors[field],
        [errors]
    );

    const set = useCallback(
        (field: string, message: string): void => {
            setErrors(
                produce(draft => {
                    draft[field] = message;
                })
            );
        },
        [setErrors]
    );

    const remove = useCallback(
        (field: string): void => {
            setErrors(
                produce(draft => {
                    delete draft[field];
                })
            );
        },
        [setErrors]
    );

    const some = useCallback(
        (): boolean => Object.values(errors).some(Boolean),
        [errors]
    );

    return useMemo(
        () => ({
            get,
            set,
            remove,
            some,
        }),
        [get, set, remove, some]
    );
};
