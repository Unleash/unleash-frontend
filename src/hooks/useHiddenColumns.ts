import { useEffect } from 'react';
import { IdType } from 'react-table';

const useHiddenColumns = (
    setHiddenColumns: <D>(param: Array<IdType<D>>) => void,
    hiddenColumns: string[],
    condition: boolean
) => {
    useEffect(() => {
        const hidden = condition ? hiddenColumns : [];
        setHiddenColumns(hidden);
    }, [setHiddenColumns, condition]);
};

export default useHiddenColumns;
