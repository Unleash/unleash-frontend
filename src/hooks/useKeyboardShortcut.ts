import { useEffect } from 'react';
import { useIsAppleDevice } from './useIsAppleDevice';

export const useKeyboardShortcut = (
    {
        key,
        modifiers = [],
        preventDefault = false,
    }: {
        key: string;
        modifiers?: Array<'ctrl' | 'alt' | 'shift'>;
        preventDefault?: boolean;
    },
    callback: () => void
) => {
    const isAppleDevice = useIsAppleDevice();
    useEffect(() => {
        const onKeyDown = (event: KeyboardEvent) => {
            if (key !== event.key) {
                return;
            }
            if (modifiers.includes('ctrl')) {
                if (isAppleDevice) {
                    if (!event.metaKey) {
                        return;
                    }
                } else {
                    if (!event.ctrlKey) {
                        return;
                    }
                }
            }
            if (modifiers.includes('alt') && !event.altKey) {
                return;
            }
            if (modifiers.includes('shift') && !event.shiftKey) {
                return;
            }
            if (preventDefault) {
                event.preventDefault();
            }

            callback();
        };

        window.addEventListener('keydown', onKeyDown);

        return () => {
            window.removeEventListener('keydown', onKeyDown);
        };
    }, [isAppleDevice, key, modifiers, preventDefault, callback]);
};
