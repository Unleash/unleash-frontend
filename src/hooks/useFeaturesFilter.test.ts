import { renderHook, act } from '@testing-library/react-hooks';
import { useFeaturesFilter } from 'hooks/useFeaturesFilter';
import { IFeatureToggle } from 'interfaces/featureToggle';
import { IConstraint, IFeatureStrategy } from 'interfaces/strategy';

test('useFeaturesFilter empty', () => {
    const { result } = renderHook(() => useFeaturesFilter([]));

    expect(result.current.filtered.length).toEqual(0);
    expect(result.current).toMatchSnapshot();
});

test('useFeaturesFilter equal', () => {
    const { result } = renderHook(() =>
        useFeaturesFilter([
            mockFeatureToggle(),
            mockFeatureToggle(),
            mockFeatureToggle(),
        ])
    );

    expect(result.current.filtered.length).toEqual(3);
    expect(result.current).toMatchSnapshot();
});

test('useFeaturesFilter project', () => {
    const { result } = renderHook(() =>
        useFeaturesFilter([
            mockFeatureToggle({ project: '1' }),
            mockFeatureToggle({ project: '2' }),
            mockFeatureToggle({ project: '2' }),
            mockFeatureToggle({ project: '3' }),
        ])
    );

    act(() => {
        result.current.setFilter({ project: '2' });
    });

    expect(result.current.filtered.length).toEqual(2);
    expect(result.current).toMatchSnapshot();
});

test('useFeaturesFilter query', () => {
    const { result } = renderHook(() =>
        useFeaturesFilter([
            mockFeatureToggle({ project: 'a' }),
            mockFeatureToggle({ project: 'ab' }),
            mockFeatureToggle({ project: 'abc' }),
            mockFeatureToggle({ project: 'abcd' }),
        ])
    );

    act(() => {
        result.current.setFilter({ project: '*', query: 'bc' });
    });

    expect(result.current.filtered.length).toEqual(2);
    expect(result.current).toMatchSnapshot();
});

test('useFeaturesFilter constraints', () => {
    const { result } = renderHook(() =>
        useFeaturesFilter([
            mockFeatureToggle({
                project: 'a',
                strategies: [
                    mockFeatureStrategy(),
                    mockFeatureStrategy(),
                    mockFeatureStrategy({
                        constraints: [
                            mockConstraint(),
                            mockConstraint({ value: 'xyz' }),
                            mockConstraint({ values: ['xyz'] }),
                        ],
                    }),
                ],
            }),
        ])
    );

    act(() => {
        result.current.setFilter({ project: '*', query: 'xyz' });
    });

    expect(result.current.filtered.length).toEqual(1);
    expect(result.current).toMatchSnapshot();
});

const mockFeatureToggle = (
    overrides?: Partial<IFeatureToggle>
): IFeatureToggle => {
    return {
        name: '1',
        description: '1',
        type: '1',
        project: '1',
        archived: false,
        enabled: false,
        stale: false,
        impressionData: false,
        strategies: [],
        variants: [],
        environments: [],
        createdAt: '22006-01-02T15:04:05Z',
        lastSeenAt: '2006-01-02T15:04:05Z',
        ...overrides,
    };
};

const mockFeatureStrategy = (
    overrides?: Partial<IFeatureStrategy>
): IFeatureStrategy => {
    return {
        id: '1',
        name: '1',
        constraints: [],
        parameters: {},
        ...overrides,
    };
};

const mockConstraint = (overrides?: Partial<IConstraint>): IConstraint => {
    return {
        contextName: '',
        operator: 'IN',
        ...overrides,
    };
};
