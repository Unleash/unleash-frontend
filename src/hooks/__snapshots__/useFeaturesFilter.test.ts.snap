// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFeaturesFilter constraints 1`] = `
Object {
  "filter": Object {
    "project": "*",
    "query": "xyz",
  },
  "filtered": Array [
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "a",
      "stale": false,
      "strategies": Array [
        Object {
          "constraints": Array [],
          "id": "1",
          "name": "1",
          "parameters": Object {},
        },
        Object {
          "constraints": Array [],
          "id": "1",
          "name": "1",
          "parameters": Object {},
        },
        Object {
          "constraints": Array [
            Object {
              "contextName": "",
              "operator": "IN",
            },
            Object {
              "contextName": "",
              "operator": "IN",
              "value": "xyz",
            },
            Object {
              "contextName": "",
              "operator": "IN",
              "values": Array [
                "xyz",
              ],
            },
          ],
          "id": "1",
          "name": "1",
          "parameters": Object {},
        },
      ],
      "type": "1",
      "variants": Array [],
    },
  ],
  "setFilter": [Function],
}
`;

exports[`useFeaturesFilter empty 1`] = `
Object {
  "filter": Object {
    "project": "*",
  },
  "filtered": Array [],
  "setFilter": [Function],
}
`;

exports[`useFeaturesFilter equal 1`] = `
Object {
  "filter": Object {
    "project": "*",
  },
  "filtered": Array [
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "1",
      "stale": false,
      "strategies": Array [],
      "type": "1",
      "variants": Array [],
    },
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "1",
      "stale": false,
      "strategies": Array [],
      "type": "1",
      "variants": Array [],
    },
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "1",
      "stale": false,
      "strategies": Array [],
      "type": "1",
      "variants": Array [],
    },
  ],
  "setFilter": [Function],
}
`;

exports[`useFeaturesFilter project 1`] = `
Object {
  "filter": Object {
    "project": "2",
  },
  "filtered": Array [
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "2",
      "stale": false,
      "strategies": Array [],
      "type": "1",
      "variants": Array [],
    },
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "2",
      "stale": false,
      "strategies": Array [],
      "type": "1",
      "variants": Array [],
    },
  ],
  "setFilter": [Function],
}
`;

exports[`useFeaturesFilter query 1`] = `
Object {
  "filter": Object {
    "project": "*",
    "query": "bc",
  },
  "filtered": Array [
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "abc",
      "stale": false,
      "strategies": Array [],
      "type": "1",
      "variants": Array [],
    },
    Object {
      "archived": false,
      "createdAt": "22006-01-02T15:04:05Z",
      "description": "1",
      "enabled": false,
      "environments": Array [],
      "impressionData": false,
      "lastSeenAt": "2006-01-02T15:04:05Z",
      "name": "1",
      "project": "abcd",
      "stale": false,
      "strategies": Array [],
      "type": "1",
      "variants": Array [],
    },
  ],
  "setFilter": [Function],
}
`;
